src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.cs

```diff
@@ -174,7 +174,18 @@ private void ReleaseConnection(SqlConnection connection)
                     return;
                 }
 
-                _connectionsPool.Push(connection);
+                try
+                {
+                    _connectionsPool.Push(connection);
+                }
+                catch
+                {
+                    // An exception (likely OutOfMemoryException) occurred while returning the connection to the pool.
```

Comment:
Why isn't the process crashing from an OutOfMemory in the first place? Surely we don't want to try hardening the *entire* process against this, since this just might move the crash somewhere else?

Reply:
It's a problem because finalizers. Finalizers almost always cause problems, but rarely provide answers.

Reply:
> but rarely provide answers.

These finalizers caught *several* issues.  They are really useful for preventing the easy mistake of forgetting to close the connection. 

That said, we could change the finalizer here to report an issue without tearing down the process.

Reply:
I guess I would have assumed the OOP would have been torn down the moment the OutOfMemoryException happened and the finalizers never would have ran; the finalizer assertion could have been kept as is. Even if we don't trigger a finalizer, the rest of the system is not safe; the Workspace has no guarantee it's handing out functional Compilations, for example...

Reply:
> I guess I would have assumed the OOP would have been torn down the moment the OutOfMemoryException happened

I don't think that actually happens.  I think OOM is a normal exception that can be recovered from.

Reply:
We frequently hit OOM cases in practice.